core/utils/tcp_flow_reconstruct.h

```diff
@@ -0,0 +1,114 @@
+#ifndef BESS_UTILS_TCP_FLOW_RECONSTRUCT_H_
+#define BESS_UTILS_TCP_FLOW_RECONSTRUCT_H_
+
+#include <vector>
+
+#include <rte_byteorder.h>
+#include <rte_config.h>
+#include <rte_errno.h>
+#include <rte_ether.h>
+#include <rte_ip.h>
+#include <netinet/tcp.h>
+
+
+#include "../packet.h"
+#include "../mem_alloc.h"
+
+namespace bess {
+namespace utils {
+
+// A utility class that accumulates TCP packet data in the correct order.
+class TcpFlowReconstruct {
+ public:
+  // Constructs a TCP flow reconstruction object that can hold initial_buflen
+  // bytes to start with.  If initial_buflen is zero, it is automatically set to
+  // 1.
+  TcpFlowReconstruct(uint32_t initial_buflen)
+      : init_seq_(0),
+        buflen_(initial_buflen),
+        received_map_(initial_buflen),
+        contiguous_len_(0) {
+    if (buflen_ <= 0) {
+      buflen_ = 1;
+      received_map_.resize(buflen_);
+    }
+
+    buf_ = (char *) mem_alloc(buflen_);
+  }
+
+  ~TcpFlowReconstruct() {
+    free(buf_);
+  }
+
+  // Returns the underlying buffer of reconstructed flow bytes.  Not guaranteed
+  // to return the same pointer between calls to InsertPacket().
+  const char *buf() const { return buf_; }
+
+  // Returns the underlying bitmap of which bytes in buf have been received.
+  // Contents are updated by each call to InsertPacket().
+  const std::vector<bool> &received_map() const { return received_map_; }
+
+  // Returns the initial data sequence number extracted from the SYN.
+  uint32_t init_seq() const { return init_seq_; }
+
+  // Returns the length of contiguous data available in the buffer starting from
+  // the beginning.  Updated every time InsertPacket() is called.
+  size_t contiguous_len() const { return contiguous_len_; }
+
+  // Adds the data of the given packet based upon its TCP sequence number.  If
+  // the packet is a SYN then we use the SYN to set the initial sequence number
+  // offset.
+  //
+  // Assumes that the IP header has no options and that the first call is the
+  // SYN.  Behavior is undefined if the first packet is not the SYN or the
+  // packet is no a TCP packet.
+  void InsertPacket(Packet *p) {
+    struct ether_hdr *eth = p->head_data<struct ether_hdr *>();
+    struct ipv4_hdr *ip = (struct ipv4_hdr *)(eth + 1);
+    struct tcphdr *tcp = (struct tcphdr *)(ip + 1);
```

Comment:
I see. If it is not too difficult, it would be better to support it though, since it should be a generic utility class (rather than solely for the milestone demo).